### ISSUE TITLE
Another implementation

### ISSUE THREAD

**Issue by smira** [stefanberger/go-kmip#5]
Hi Stefan!

Several days ago (exactly at the same time as you), I've started implementation of KMIP library: https://github.com/smira/go-kmip

I wouldn't have started it if I knew there's another one, but looks like we did it almost simultaneously.

**Comment by stefanberger** [comment]
Oh my...

**Comment by stefanberger** [comment]
What now ?

**Comment by stefanberger** [comment]
I think your annotations will produce a lot less code in the long run...

**Comment by stefanberger** [comment]
What I would like to get to is something like this here: https://github.com/stefanberger/go-jose/commit/604a4c6fcdffd3c70eb1d94aaa6570bdbb9588dd

**Comment by smira** [comment]
so far have only encoding/decoding implemented, and I was going to move from that to some simple client/server protocol.

all I need to implement for my private project is DiscoverVersion, Get, Create, GetAttribute(s), Delete.

**Comment by stefanberger** [comment]
I initially only wanted 'Get'. But that 'Get' may not always work, though it seems to work for pykmip. I don't think an HSM would want to give me the private key, so pykmip may be special in this respect. I'll have to see once I use some better server. That's why I added the private key operations like signing and decrypting assuming that a server will keep the private key private. Then I needed to get the public key of a private key using the uuid of the private key. There should be a link object from the private key to the public key, but it doesn't seem to be there in pykmip. This made me add the GetAttributes function. Well, so much for just implementing 'Get'...

**Comment by smira** [comment]
I actually need to implement KMIP server, not client in my case which will only support symmetric key management.

**Comment by stefanberger** [comment]
Any way we could join forces ?

**Comment by smira** [comment]
I think it makes sense to have a common layer for encoding/decoding stuff?

**Comment by stefanberger** [comment]
I agree. And build objects representing keys on top of that.

**Comment by stefanberger** [comment]
Would you want to host a file like this one: https://github.com/stefanberger/go-kmip/blob/master/kmip/objects.go

And the client tools I have in the tools directory? https://github.com/stefanberger/go-kmip/tree/master/tools

**Comment by stefanberger** [comment]
Also, what do you intend to do with it long-term after you have implemented the commands you want. My interest is mostly on the client side...

**Comment by smira** [comment]
So far I don't have real long-term plan, as I don't have even the first revision complete. 

My next step is to implement two high-level entities: `Client` and `Server` which allow client-server interaction with same basic requests, but not doing any meaningful work, and package customers can implement their own business logic around it.

**Comment by stefanberger** [comment]
I just found yet another implementation that doesn't look very complete, either: https://github.com/gemalto/kmip-go

**Comment by smira** [comment]
Yep, looks like there're three now. I've managed to complete the core server/client implementation. 

Implementation is now fairly complete, all I have for this common library is to add structs for specific KMIP requests/response payloads I need.

**Comment by stefanberger** [comment]
@smira You'll accept patches extending your go-kmip, right?
